torch/_refs/__init__.py

```diff
@@ -2775,22 +2776,25 @@ def stack(tensors: TensorSequenceType, dim: int = 0) -> TensorLikeType:
     return torch.cat([t.unsqueeze(wrapped_dim) for t in tensors], dim)
 
 
+# CompositeImplicitAutograd - don't register decomp
 @out_wrapper()
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT,
+)
 def softmax(
     a: TensorLikeType,
     dim: int,
-    *,
     dtype: Optional[torch.dtype] = None,
 ) -> TensorLikeType:
     result_dtype = dtype or a.dtype
-    computation_dtype = utils.get_computation_dtype(a.dtype)
-    a_ = _maybe_convert_to_dtype(a, computation_dtype)
-    assert isinstance(a_, TensorLike)  # to avoid MyPy error for amax
-    a_max = amax(a_, dim, keepdim=True)
-    a_exp = exp(a_ - a_max)
-    return _maybe_convert_to_dtype(
-        true_divide(a_exp, sum(a_exp, dim, keepdim=True)), result_dtype
-    )  # type: ignore[return-value]
+    # amax is to improve FP precision with large inputs, but sub doesn't work
+    # for bool and amax fails for complex.  However, these dtypes are not
+    # supported by kernels anyway at the moment.
+    m = torch.amax(a, dim, keepdim=True)
+    e = torch.exp(a - m)
+    r = e / torch.sum(e, dim, keepdim=True)
+    return _maybe_convert_to_dtype(r, result_dtype)  # type: ignore[return-value]
```

Comment:
`dtype` arg wasn't correctly handled previously, and is still not correctly handled. If `dtype` is supplied, `a` is first converted to dtype.

Reply:
pushed a version that can use dtype before converting to comp dtype and when converting to result dtype